weave/trace/annotation_parser.py

```diff
@@ -0,0 +1,151 @@
+from __future__ import annotations
+
+import re
+from inspect import Signature
+
+from pydantic import BaseModel
+
+CONTENT_CLASS_NAME = "weave.type_wrappers.Content.content.Content"
+
+
+class ContentAnnotation(BaseModel):
+    base_type: str
+    media_class: str
+    raw_annotation: str
+    type_hint: str | None = None
+
+
+def parse_data_annotation(annotation_string: str) -> ContentAnnotation | None:
+    """
+    The function expects the string to be of the form:
+    typing.Annotated[<SomeType>, weave.type_wrappers.Content.content.Content[typing.Literal[<format>]]]
+
+    where <format> is a supported media format such as mp3 or wav and <SomeType> is a str (optionally base64) or bytes.
+
+    Args:
+        annotation_string: The string representation of the type annotation.
+
+    Returns:
+        ContentAnnotation | None: An instance of ContentAnnotation if the parsing is successful,
+    """
+    # Content annotation with a format specifier
+    # Example: "typing.Annotated[typing.Union[str, bytes], weave.type_wrappers.Content.Content.Content[typing.Literal['mp3']]]"
+    # Explanation:
+    # - r"typing\.Annotated\[": Matches the literal "typing.Annotated["
+    # - r"(.+?)": Group 1. Captures the base type (non-greedy match for any characters).
+    #            This allows capturing complex types like `typing.Union[str, bytes]`.
+    # - r",\s*": Matches a comma followed by optional whitespace (this comma separates base_type from the Content annotation).
+    # - fr"{re.escape(content_class_name)}\[": Matches the content class name followed by "[".
+    # - r"typing\.Literal\[": Matches "typing.Literal[".
+    # - r"['\"](.+?)['\"]": Group 2. Captures any literal (non-greedy).
+    # - r"\]\]": Matches the two closing square brackets for Literal and Content.
+    # - r"\]": Matches the final closing square bracket for Annotated.
+
+    pattern_with_format = re.compile(
+        r"typing\.Annotated\["
+        r"(.+?),\s*"  # Group 1: Base type (non-greedy)
+        rf"{re.escape(CONTENT_CLASS_NAME)}\["
+        r"typing\.Literal\[['\"](.+?)['\"]\]"  # Group 2: Any literal, non-greedy
+        r"\]"  # Closing bracket for Content[...]
+        r"\]"  # Closing bracket for Annotated[...]
+    )
```

Comment:
This should happen outside of the function body so you don't re-compile on each invocation